linker/shared/src/main/scala/org/scalajs/linker/frontend/optimizer/IntegerDivisions.scala

```diff
@@ -0,0 +1,693 @@
+import org.scalajs.ir.Types._
+import org.scalajs.ir.WellKnownNames._
+
+import org.scalajs.linker.backend.emitter.LongImpl
+
+/** Rewrites for integer divisions by constants using adds, muls and shifts.
+ *
+ *  This class uses strategies from Hacker's Delight, Chapter 10.
+ */
+private[optimizer] final class IntegerDivisions(
+    useRuntimeLong: Boolean, isWebAssembly: Boolean) {
+
+  import IntegerDivisions._
+
+  /** Should we apply the optimized rewrite for division by a constant?
+   *
+   *  The particular conditions were decided based on the results of
+   *  microbenchmarks. We apply the rewrite if the result was measured to be
+   *  reliably at least as good as the division.
+   *
+   *  Currently, we apply the rewrite if at least one of the following
+   *  conditions holds:
+   *
+   *  - the divisor is zero (rewrite to a trivial `throw`);
+   *  - the divisor is a power of 2 (including negative values for the signed
+   *    operations, and including 1's);
+   *  - it is a `long` operation and we are targeting JavaScript; or
+   *  - it is an `int` operation and we are targeting WebAssembly.
+   *
+   *  The remaining possibilities were measured to be detrimental. The
+   *  measurements actually make sense. I had even predicted them before
+   *  running the benchmarks.
+   *
+   *  For non-powers of two, rewriting `int` operations on JavaScript is
+   *  detrimental, as we don't have efficient access to the hi result of a
+   *  32-bit multiplication. We don't have it either for 64 bits, but `long`
+   *  division is so bad on JS that it is a win anyway.
+   *
+   *  On WebAssembly, we have efficient access to the hi result of a 32-bit
+   *  multiplication, but not for 64 bits.
+   */
+  def shouldRewriteDivision[T](op: BinaryOp.Code, divisor: T)(
+      implicit int: UnsignedIntegral[T]): Boolean = {
+
+    val (isSigned, isLong) = op match {
+      case BinaryOp.Int_/ | BinaryOp.Int_%                     => (true, false)
+      case BinaryOp.Long_/ | BinaryOp.Long_%                   => (true, true)
+      case BinaryOp.Int_unsigned_/ | BinaryOp.Int_unsigned_%   => (false, false)
+      case BinaryOp.Long_unsigned_/ | BinaryOp.Long_unsigned_% => (false, true)
+    }
+
+    val negativeDivisor = isSigned && int.lt(divisor, int.zero)
+    val absDivisor = if (negativeDivisor) int.negate(divisor) else divisor
+
+    (
+      int.isUnsignedPowerOf2OrZero(absDivisor) ||
+      isLong != isWebAssembly
+    )
+  }
+
+  /** Generates optimized code for division by a given constant divisor.
+   *
+   *  The generated code must be integrated in a context where the numerator is
+   *  bound to the local name `IntegerDivisions.NumeratorArgName` (as if it had
+   *  been inlined from a method taking `NumeratorArgName` as formal parameter).
+   *
+   *  This method produces correct code for all divisors (including zero).
+   *  However, for best performance, it should only be used if
+   *  `shouldRewriteDivision` returns `true`.
+   */
```

Comment:
Maybe drop a line that the code is intended to be folded by the optimizer after?

IIUC otherwise we'll generate unnecessary shifts when `data.shift == 0`. Which, IIUC, can happen at least for division by 3.